Compile Kotlin DSL scripts against the public ABI jar from the distribution

jbartok opened the pull request:
* Fix https://github.com/gradle/gradle/issues/26484
* Relates to #15886
----

This PR does a few main things:
* build a new Public API ABI jar (`gradle-public-api-legacy`), with dependencies equivalent to the current generated API jar
* embed that Public API ABI jar in the distribution (~6MiB)
* let the Kotlin DSL script compilation use the ABI JAR instead of requiring the Gradle API jar generation
* let the TAPI model builders use the ABI JAR instead of requiring the Gradle API jar generation

It doesn't change:
* the fact that the ABI jar contains internals
* the fact that plugin projects with a dependency on `gradleKotlinDsl()` (or applying `kotlin-dsl` plugin) use the generated API JAR

The first few commits are opportunistic refinements.

A side effect of this PR is that our integration tests that rely on `:distribution-full` now don't need to re-generate the API jar on every production code change when running tests that use Kotlin DSL scripts. They get faster to start. It is not the case for integ tests using the other smaller testing distributions as they don't get a ABI jar.

Potential breaking changes:
- scripts importing from relocated dependencies `org.gradle.internal.impldep.*` will not compile anymore

Potential follow ups:
- let testing distributions (basics, core, jvm ...) embed an adequate ABI jar in order to remove the fallback to the generated API jar (simplifying production code and making more of our integ tests faster)

----

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

eskatos committed:
Polish :test-kit:generateTestKitPackageList

Properly wiring resource generation to the main source set

eskatos committed:
Add test cases for constants inlining in :java-api-extractor

eskatos committed:
Remove org.gradle.kotlin.dsl from gradle-public-api-internal

eskatos committed:
Add missing public API modules to gradle-public-api-internal

eskatos committed:
Add gradle-public-api-internal to the full distribution

Simplify gradlebuild.distributions.gradle.kts by use configuration factories and removing Kotlin delegates

eskatos committed:
Move Java Toolchains integration tests to :toolchains-jvm

eskatos committed:
Kotlin DSL script compilation uses gradle-public-api-internal JAR from the distro

eskatos committed:
Remove spurious import from snippet script

eskatos committed:
Add test coverage for using toolchainManagement { jvm {} } without any jvm toolchain plugin applied

eskatos committed:
Let Kotlin DSL TAPI builders use the script compilation classpath

eskatos commented:
@bot-gradle test RFN without PTS

bot-gradle commented:
The following builds have failed:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_ReadyforNightly_Trigger/105961693)
  - [Build Scan](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.tags=CI&amp;search.names=gitCommitId&amp;search.values=8c0362897548b0c12288655014ec613c925703ff&amp;)

eskatos commented:
Running local benchmarks again (`android-100-kts` build logic change IDE sync with IP), this time measuring the `CompileScriptBuildOperationType` build operation, I can see that script compilation itself is ~35% faster 🚀 

EDIT: this number is outdated and was measured at a stage where the changeset lacked some correctness, see below for up-to-date numbers

ljacomet commented on build-logic/packaging/src/main/kotlin/gradlebuild.distributions.gradle.kts:
```diff
@@ -351,6 +359,18 @@ fun libraryResolver(extends: List<Configuration>) =
+            attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.LIBRARY))
+            attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named(LibraryElements.JAR))
+        }
+        isCanBeResolved = true
```
🤔 Should we start using `resolvable` to not have to specify those?

ljacomet commented on build-logic/packaging/src/main/kotlin/gradlebuild.distributions.gradle.kts:
```diff
@@ -83,19 +83,27 @@ val pluginsRuntimeOnly by bucket()
-val agentsRuntimeClasspath by libraryResolver(listOf(agentsRuntimeOnly))
+val agentsRuntimeClasspath by runtimeLibraryResolver(listOf(agentsRuntimeOnly))
 agentsRuntimeClasspath.description = "Resolves to all Jars that need to be added as agents"
+val gradlePublicApiRuntimeClasspath by apiLibraryResolver(listOf(publicApiOnly)) // TODO: name -> compile
```
❌ Let's fix that TODO and not have runtime in the name of something that is a compile classpath

ljacomet commented on packaging/public-api/build.gradle.kts:
```diff
@@ -28,19 +31,50 @@ dependencies {
 }
 
+// The JAR built by this module includes the ABI of internals
+val internalBaseName = gradleModule.identity.baseName.map { "$it-internal" }
```
❌ We should not change the coordinates of what we publish. I motivated this choice [here](https://github.com/gradle/gradle/issues/29483#issuecomment-2791668178)

ljacomet commented on packaging/public-api/build.gradle.kts:
```diff
@@ -28,19 +31,50 @@ dependencies {
     externalApi(libs.groovy)
+    // Groovy modules that are part of our API
+    // See DependencyClassPathProvider and DependencyClassPathNotationConverter
+    externalApi(libs.groovyAnt)
```
❓ Since we are expanding on dependencies, maybe consider [this comment](https://github.com/gradle/gradle/issues/29483#issuecomment-3078805705)

eskatos replied:
As discussed with Louis, I will restore the `-internal` variant as it was and introduce a new one, not-published, with the complete list of dependencies that the current generated API jar has in order to limit the potential for this PR breaking existing builds.

eskatos replied:
Done

eskatos committed:
Rename ModuleJar to GradleModuleJar for clarity

eskatos commented on build-logic-commons/module-identity/src/main/kotlin/gradlebuild/jar/ModuleJar.kt:
Renamed to `GradleModuleJar` for clarity

eskatos committed:
Refine kotlin-dsl/build.gradle.kts by removing the apiStubElements variant instead of clearing its outgoing artifacts

eskatos committed:
Restore content of gradle-public-api-internal

eskatos committed:
Add second -legacy variant to :public-api matching the current content of gradleApi()

jvandort commented on platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinDslScriptCompilationIntegrationTest.kt:
```diff
@@ -0,0 +1,52 @@
+
+        // and:
+        executer.requireOwnGradleUserHomeDir("Testing API jar generation")
+        val generatedJarsDirectory = "user-home/caches/${this.distribution.version.version}/generated-gradle-jars"
```
Can we somehow link this test to the logic that calculates this cache dir (maybe by referencing [this constant](https://github.com/gradle/gradle/blob/a23d70ba1e9c7566dce16aa84da6a2f8df553629/subprojects/core/src/main/java/org/gradle/cache/internal/GeneratedGradleJarCache.java#L31))? If the cache dir changes, or we remove the generated jar functionality (sometime in the future) this test will still happily pass

jvandort commented on platforms/extensibility/test-kit/build.gradle.kts:
```diff
@@ -53,14 +53,13 @@ dependencies {
 val generateTestKitPackageList by tasks.registering(PackageListGenerator::class) {
-    classpath.from(sourceSets.main.map { it.runtimeClasspath })
-    outputFile = layout.buildDirectory.file("runtime-api-info/test-kit-relocated.txt")
+    classpath.from(configurations.runtimeClasspath.get().elements)
```
`.elements` should not be necessary here. A configuration is already a FileCollection

jvandort commented on platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt:
```diff
@@ -64,34 +67,58 @@ fun kotlinScriptClassPathProviderOf(project: Project) =
+                    false -> cp
+                }
+            }
+        // If the distro doesn't contain the ABI jar, fallback to the generated API jar
```
Eventually it might make sense for there not to be one gradle API jar, but for each distribution to have their own API jar that it generates and embeds on its own

jvandort replied:
I see this is mentioned in the PR description

jvandort commented on platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt:
```diff
@@ -64,34 +67,58 @@ fun kotlinScriptClassPathProviderOf(project: Project) =
+        moduleRegistry.findModule("gradle-public-api-legacy")
+            ?.allRequiredModulesClasspath
+            ?.asFiles
+            ?.let { files -> DefaultClassPath.of(files) }
```
allRequiredModulesClasspath is already a `ClassPath`, why do we have this logic?

jvandort commented on packaging/public-api/build.gradle.kts:
```diff
@@ -30,15 +31,76 @@ dependencies {
+    legacyGradleApiElements {
+        outgoing {
+            capability(legacyBaseName.map { "$group:$it:$version" })
+        }
```
How come we're configuring all of this in the buildscript? It seems like most if not all of this logic could be moved to the plugin and configured where the tasks/configurations are created

eskatos replied:
Right! Moved to the plugin except the dependency declarations.

jvandort commented on build-logic/packaging/src/main/kotlin/gradlebuild.distributions.gradle.kts:
```diff
@@ -333,91 +374,91 @@ fun generatedPropertiesFileFor(name: String) =
-fun consumableSourcesVariant(name: String, extends: List<Configuration>, vararg artifacts: Any) =
-    configurations.create("${name}Elements") {
+fun consumableSourcesVariant(name: String, extends: List<NamedDomainObjectProvider<DependencyScopeConfiguration>>, sourceDir: Provider<Directory>) =
+    configurations.register("${name}Elements") {
```
Any reason we didn't change these over to `consumable`?

jvandort reviewed (commented):
Got through most of this but got pulled away to work on another issue. Have some thoughts though

eskatos committed:
Always use GeneratedGradleJarCache.CACHE_KEY to reference the path of the generated jars

eskatos commented on build-logic/packaging/src/test/kotlin/gradlebuild/packaging/GradleDistributionInstallTest.kt:
```diff
@@ -226,6 +226,7 @@ class GradleDistributionInstallTest {
         val coreRuntimeClasspath by configurations.creating
         val runtimeClasspath by configurations.creating
         val agentsRuntimeClasspath by configurations.creating
+        val gradlePublicAbiClasspath by configurations.creating
```
The same as the other lines above. It creates the configurations that are referenced by the distribution copy spec.

eskatos committed:
Move Gradle Module JAR manifest configuration to module-identity plugin

eskatos committed:
Move JetBrains Annotations out of the public api jar to the legacy api jar

eskatos committed:
Consolidate public API module-properties generation and variant naming

eskatos committed:
Remove unnecessary ClassPath => Files => ClassPath creation

lkasso committed:
add deprecation docs

lkasso commented:
I added a deprecation item in the upgrade_guide.adoc but feel free to update/remove if you think it's not warranted.

lkasso reviewed (approved):
Docs approved.

eskatos commented on platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt:
```diff
@@ -64,34 +67,58 @@ fun kotlinScriptClassPathProviderOf(project: Project) =
+                // Declarative Gradle is not part of the public api yet
+                // Only add it to the compilation classpath if enabled
+                when (gradleProperties.isDclEnabled) {
+                    true -> cp + moduleRegistry.getModule("gradle-project-features").classpath
```
You're right, these classes were never in the generated `gradle-api.jar` (they're not in `PublicApi.includes`). The conditional is here because with Kotlin-DSL-support-for-DCL requires `org.gradle.features.*` to be available. Without this, builds enabling that support would fail to compile against the new ABI path. Master got away with it because its compilation classpath came from `ClassPathRegistry.GRADLE_API` (the in-distribution registry view), which happens to include `gradle-project-features-*.jar`.

The cleanest fix is to graduate the relevant `org.gradle.features.*` packages to public API; which would put them in the legacy ABI jar automatically and drop the conditional. But we are not ready for it. I'll add a clear TODO for the declarative team here.

eskatos commented:
New measurements
* [CI regression tests](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_PerformanceTestSlowLinux_Trigger/113112972?buildTab=report_project104067_Performance)
    * `buildSrc abi change | largeJavaMultiProjectKotlinDsl | BuildSrcApiChangePerformanceTest` is ~14% faster (overall build time)
    * `first use | largeJavaMultiProjectKotlinDsl | JavaFirstUsePerformanceTest` is ~6% faster (overall build time)
* Local benchmarks with [IP testbed projects](https://github.com/gradle/isolated-projects-testbed)
    * IDE Sync with build-logic ABI change, Parallel IP on `android100Kts` (`resync_bl_abi_change_pip_warm`) is ~14% faster (overall sync time) [benchmark.html.zip](https://github.com/user-attachments/files/28277005/benchmark.html.zip)

eskatos commented:
@bot-gradle test RFN

bot-gradle commented:
<!-- {"replyTargetCommentId":4553003942,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_Stage_ReadyforNightly_Trigger","buildId":"113152363","headRef":"b09573c9f1776449f7d6812882ede80f42171666"}]} -->
I've triggered the following builds:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/build/113152363)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=b09573c9f1776449f7d6812882ede80f42171666,ReadyforNightly&amp;)

bot-gradle commented:
The following builds have passed:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_ReadyforNightly_Trigger/113152363)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=b09573c9f1776449f7d6812882ede80f42171666,ReadyforNightly&amp;)

jbartok commented on platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_9.adoc:
```diff
@@ -36,6 +36,39 @@ The previous step will help you identify potential problems by issuing deprecati
+[[kotlin_dsl_impldep_compilation]]
+==== Kotlin DSL scripts can no longer reference relocated internal dependencies
+
+Kotlin DSL build scripts (`.gradle.kts`) are now compiled against a pre-built public API ABI JAR from the Gradle distribution instead of the dynamically generated Gradle API JAR.
```
```suggestion
Kotlin DSL build scripts (`.gradle.kts`) are now compiled against a pre-built public API ABI JAR from the Gradle distribution instead of the dynamically generated Gradle API JAR (both jars that contain only the shape of classes - what callers compile against - but not implementations; so class names, public fields and method and so on, but no method bodies, private fields and such).
```

eskatos replied:
The generated API jar is not stripped of implementations, it is an actually usable at runtime JAR. Only the ABI jar is stripped. I think with the other documentation edits, this is addressed.

jbartok commented on platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_9.adoc:
```diff
@@ -36,6 +36,39 @@ The previous step will help you identify potential problems by issuing deprecati
+// import com.google.gson.Gson
+----
+
+NOTE: This change does not affect plugin projects that depend on `gradleApi()` or `gradleKotlinDsl()`. Those continue to use the full generated Gradle API JAR as before.
```
```suggestion
NOTE: This change does not affect plugin projects that have `gradleApi()` or `gradleKotlinDsl()` among their dependencies. Those continue to use the full generated Gradle API JAR as before.
```

jbartok commented on build-logic-commons/basics/src/main/kotlin/gradlebuild/basics/PublicApiVariants.kt:
```diff
@@ -0,0 +1,16 @@
+object PublicApiVariants {
+    const val INTERNAL_SUFFIX = "-internal"
+    const val LEGACY_SUFFIX = "-legacy"
+    const val LEGACY_MODULE_NAME = "gradle-public-api$LEGACY_SUFFIX"
```
This to me is very confusing: `gradle-public-api-legacy.jar` referts to a NEWLY ADDED, published ABI jar, so why use the words "legacy" and "api"?

eskatos replied:
I added clarifying comments in 683208b40344a65ece5455c21aa094b86d1d4da3

jbartok commented on testing/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegritySpec.groovy:
```diff
@@ -57,11 +57,15 @@ class DistributionIntegritySpec extends DistributionIntegrationSpec {
         def jarsWithDuplicateFiles = [:]
         def classesIndex = [:] as HashMap<String, List<String>> // class name -> list of containing jars
         jars.each { jar ->
+            // The API/ABI jars purposely duplicate class entries that also live in other distro jars.
```
I think we should use the two terms of "API" or "ABI" either very precisely or just stick to one of them. The changes feel a bit like "using both, but it's limping a bit". And I'm worried about how many people do clearly know the distinction between them... I would suggest to give up using one in names and such, maybe stick with API only and then only talk about ABI when the difference between the two is the main topic.

eskatos replied:
The distinction is important. But I agree this was not conveyed uniformly throughout the code. I adjusted things to clarify this in 683208b40344a65ece5455c21aa094b86d1d4da3

jbartok reviewed (commented):
Looks good to me, just made some documentation/naming related comments.

eskatos committed:
Update platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_9.adoc

Co-authored-by: József Bartók <jbartok@gradle.com>
Signed-off-by: Paul Merlin <paul@nosphere.org>

eskatos committed:
Clarify the "legacy" and ABI terminology for the public API jars

These terms confused reviewers: "legacy" on a newly added artifact, and
"API" and "ABI" used interchangeably for the same jar. Define them once
where the variant names live and use them precisely, so the intent (stubs
to compile against, internals still bundled, an internals free variant
planned) stays discoverable in code and upgrade notes without renaming
published artifacts.

eskatos commented:
A few notes about the effects of this PR.


|                                                            | Size (MiB) | `.class` files |
| ---------------------------------------------------------- | ---------: | -------------: |
| **OLD** `gradle-api.jar` (generated fat shaded jar)        | **181.84** |     **45,801** |
| **NEW** ABI classpath (legacy ABI jar + 21 transitive deps) |  **24.30** |     **22,929** |
| **Reduction**                                              |    −157.54 |        −22,872 |
| **Reduction %**                                            |  **−86.6%** |     **−49.9%** |



- The legacy ABI jar contains **method signatures only**, no bodies → small per-class.
- The transitive deps (`kotlin-stdlib`, `groovy`, `ant`, …) are **full implementations** → normal-sized per-class.
- So new-classpath classes are a mix: cheap ABI stubs from the legacy jar (~12,800 classes) plus full classes from the deps (~10,100). The old classpath was 45,801 classes, all full implementations, all relocated into one big shaded jar.

jbartok commented:
Approved!

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"triggeredTeamCityBuilds":[]} -->
WARN: Based on labels, this pull request addresses notable issue but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"triggeredTeamCityBuilds":[]} -->
WARN: Based on labels, this pull request addresses notable issue but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"triggeredTeamCityBuilds":[]} -->
WARN: Based on labels, this pull request addresses notable issue but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.

eskatos commented:
@bot-gradle test RFN without PTS

bot-gradle commented:
<!-- {"replyTargetCommentId":4565026821,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_Stage_ReadyforNightly_Trigger","buildId":"113219381","headRef":"683208b40344a65ece5455c21aa094b86d1d4da3"}]} -->
I've triggered the following builds:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/build/113219381)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=683208b40344a65ece5455c21aa094b86d1d4da3,ReadyforNightly&amp;)

bot-gradle commented:
The following builds have failed:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_ReadyforNightly_Trigger/113219381)
  - [Build Scan](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.tags=CI&amp;search.names=gitCommitId&amp;search.values=683208b40344a65ece5455c21aa094b86d1d4da3&amp;)

eskatos commented:
@bot-gradle test RFN without PTS

bot-gradle commented:
<!-- {"replyTargetCommentId":4567633276,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_Stage_ReadyforNightly_Trigger","buildId":"113229821","headRef":"0f27bfcf37dd599ab2eadb982d6a3a9d9fd86b85"}]} -->
I've triggered the following builds:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/build/113229821)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=0f27bfcf37dd599ab2eadb982d6a3a9d9fd86b85,ReadyforNightly&amp;)

bot-gradle commented:
The following builds have passed:
- [Ready for Nightly (Trigger) (Check)](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_ReadyforNightly_Trigger/113229821)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=0f27bfcf37dd599ab2eadb982d6a3a9d9fd86b85,ReadyforNightly&amp;)

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"triggeredTeamCityBuilds":[]} -->
WARN: Based on labels, this pull request addresses notable issue but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"triggeredTeamCityBuilds":[]} -->
WARN: Based on labels, this pull request addresses notable issue but no changes to [release note](https://github.com/gradle/gradle/blob/master/platforms/documentation/docs/src/docs/release/notes.md) found.
